File: maxrmac/utils/solve_tone.m
function [f, b, Rxx, E, state_out] = solve_tone(H, theta, w, Lxu, idx_start, idx_end, state_in)
    %SOLVE_TONE Per-tone Lagrangian optimization for maxRMAC
    %   [f, b, Rxx, E, state_out] = SOLVE_TONE(H, theta, w, Lxu, ...)
    %   solves the per-tone optimization for rate maximization subject to
    %   weighted energy penalty. Uses L-BFGS over Cholesky factors.
    %
    %   Maximizes: sum_u delta(u) * log|S_u| - sum_u w(u) * trace(Rxx_u)
    %   where S_u = I + sum_{j=1}^u H_j * Rxx_j * H_j' and delta = theta - theta_{next}
    %
    %   Inputs:
    %       H           Ly-by-Ltot channel matrix for this tone
    %       theta       U-by-1 rate weights
    %       w           U-by-1 energy weights (Lagrange multipliers)
    %       Lxu         1-by-U antennas per user
    %       idx_start   1-by-U starting antenna indices
    %       idx_end     1-by-U ending antenna indices
    %       state_in    (optional) warm-start state from previous solve
    %
    %   Outputs:
    %       f           scalar Lagrangian value
    %       b           U-by-1 per-user rates in nats
    %       Rxx         U-by-1 cell array of covariance matrices
    %       E           U-by-1 per-user energies
    %       state_out   state structure for warm-starting

    [Ly, ~] = size(H);
    U = length(Lxu);

    % sort users by decreasing theta for polymatroid vertex
    [stheta, order] = sort(theta, 'descend');
    sw = w(order);
    sLxu = Lxu(order);

    % build per-user channel blocks in sorted order
    H_blocks = cell(U, 1);

    for u = 1:U
        orig = order(u);
        cols = idx_start(orig):idx_end(orig);
        H_blocks{u} = H(:, cols);
    end

    % coefficients: delta_k = theta_k - theta_{k+1}, theta_{U+1} = 0
    delta = stheta - [stheta(2:U); 0];

    % process warm-start state
    state_valid = false;

    if nargin >= 7 && ~isempty(state_in)

        if isstruct(state_in) && isfield(state_in, 'order') && isequal(state_in.order(:), order(:))

            if isfield(state_in, 'block_sizes') && isequal(state_in.block_sizes(:), sLxu(:))
                state_valid = true;
            end

        end

    end

    if ~state_valid
        state_in = [];
    end

    % run l-bfgs optimization
    [R_blocks, state_internal] = maximize_lagrangian_lbfgs(H_blocks, delta, sw, sLxu, Ly, state_in);

    % compute achieved rates (nats) for sorted users
    b_sorted = zeros(U, 1);
    S_prev = eye(Ly);

    for u = 1:U
        S_curr = S_prev + H_blocks{u} * R_blocks{u} * H_blocks{u}';
        S_curr = 0.5 * (S_curr + S_curr');
        b_sorted(u) = real(logdet_spd(S_curr) - logdet_spd(S_prev));
        S_prev = S_curr;
    end

    % compute energies for sorted users
    E_sorted = zeros(U, 1);

    for u = 1:U
        E_sorted(u) = real(trace(R_blocks{u}));
    end

    % map back to original ordering
    b = zeros(U, 1);
    E = zeros(U, 1);
    Rxx = cell(U, 1);
    b(order) = b_sorted;
    E(order) = E_sorted;

    for u = 1:U
        orig = order(u);
        Rxx{orig} = R_blocks{u};
    end

    % compute Lagrangian value
    f = theta' * b - w' * E;

    % prepare output state
    if nargout >= 5
        state_out = state_internal;
        state_out.order = order;
        state_out.block_sizes = sLxu;
    else
        state_out = [];
    end

end

function [R_blocks, state_out] = maximize_lagrangian_lbfgs(H_blocks, delta, w, block_sizes, Ly, state_in)
    %MAXIMIZE_LAGRANGIAN_LBFGS L-BFGS solver for per-tone Lagrangian

    % l-bfgs parameters
    max_iter = 200;
    tol_grad = 1e-6;
    m_hist = 12;
    step_init = 1.0;
    step_min = 1e-12;
    beta = 0.5;
    armijo_c = 1e-4;
    max_linesearch = 24;

    U = numel(H_blocks);

    % initialize cholesky factors
    B_blocks = cell(U, 1);
    S_hist = {};
    Y_hist = {};

    if ~isempty(state_in) && isfield(state_in, 'B_blocks')
        valid = true;

        for u = 1:U
            Bu = state_in.B_blocks{u};

            if isempty(Bu) || any(size(Bu) ~= block_sizes(u))
                valid = false;
                break;
            end

        end

        if valid
            B_blocks = state_in.B_blocks;

            if isfield(state_in, 'S_hist')
                S_hist = state_in.S_hist;
            end

            if isfield(state_in, 'Y_hist')
                Y_hist = state_in.Y_hist;
            end

        else

            for u = 1:U
                B_blocks{u} = sqrt(1e-6) * eye(block_sizes(u));
            end

        end

    else

        for u = 1:U
            B_blocks{u} = sqrt(1e-6) * eye(block_sizes(u));
        end

    end

    x = pack_blocks(B_blocks, block_sizes);
    [phi, grad, R_blocks] = eval_objective(x, H_blocks, delta, w, block_sizes, Ly);

    if isinf(phi)
        % reset to safe initial point

        for u = 1:U
            B_blocks{u} = sqrt(1e-6) * eye(block_sizes(u));
        end

        x = pack_blocks(B_blocks, block_sizes);
        S_hist = {};
        Y_hist = {};
        [phi, grad, R_blocks] = eval_objective(x, H_blocks, delta, w, block_sizes, Ly);
    end

    % main l-bfgs loop
    for iter = 1:max_iter

        if norm(grad) < tol_grad
            break;
        end

        dir = lbfgs_direction(grad, S_hist, Y_hist);
        dir_deriv = real(grad' * dir);

        if dir_deriv >= 0
            dir = -grad;
            dir_deriv = real(grad' * dir);
            S_hist = {};
            Y_hist = {};
        end

        step = step_init;
        accepted = false;
        x_new = x;
        phi_new = phi;
        grad_new = grad;
        R_new = R_blocks;

        for ls = 1:max_linesearch
            x_trial = x + step * dir;
            [phi_trial, grad_trial, R_trial] = eval_objective(x_trial, H_blocks, delta, w, block_sizes, Ly);

            if ~isinf(phi_trial) && phi_trial <= phi + armijo_c * step * dir_deriv
                accepted = true;
                x_new = x_trial;
                phi_new = phi_trial;
                grad_new = grad_trial;
                R_new = R_trial;
                break;
            end

            step = step * beta;

            if step < step_min
                break;
            end

        end

        if ~accepted
            % fall back to gradient descent
            dir = -grad;
            dir_deriv = real(grad' * dir);
            step = step_init;

            for ls = 1:max_linesearch
                x_trial = x + step * dir;
                [phi_trial, grad_trial, R_trial] = eval_objective(x_trial, H_blocks, delta, w, block_sizes, Ly);

                if ~isinf(phi_trial) && phi_trial <= phi + armijo_c * step * dir_deriv
                    accepted = true;
                    x_new = x_trial;
                    phi_new = phi_trial;
                    grad_new = grad_trial;
                    R_new = R_trial;
                    break;
                end

                step = step * beta;

                if step < step_min
                    break;
                end

            end

            if ~accepted
                break;
            end

            S_hist = {};
            Y_hist = {};
        end

        s = x_new - x;
        y = grad_new - grad;
        sy = real(s' * y);

        if sy > 1e-12

            if length(S_hist) >= m_hist
                S_hist = S_hist(2:end);
                Y_hist = Y_hist(2:end);
            end

            S_hist{end + 1} = s;
            Y_hist{end + 1} = y;

        else
            S_hist = {};
            Y_hist = {};
        end

        x = x_new;
        phi = phi_new;
        grad = grad_new;
        R_blocks = R_new;
    end

    % extract final R blocks from B blocks
    B_blocks = unpack_blocks(x, block_sizes);

    for u = 1:U
        Ru = B_blocks{u} * B_blocks{u}';
        R_blocks{u} = 0.5 * (Ru + Ru');
    end

    state_out = struct();
    state_out.B_blocks = B_blocks;
    state_out.S_hist = S_hist;
    state_out.Y_hist = Y_hist;

end

function [phi, grad_vec, R_blocks] = eval_objective(x, H_blocks, delta, w, block_sizes, Ly)
    %EVAL_OBJECTIVE Evaluate objective and gradient

    B_blocks = unpack_blocks(x, block_sizes);
    U = numel(H_blocks);
    eye_Ly = eye(Ly);
    S = eye_Ly;
    logdet_S = zeros(U, 1);
    invS = cell(U, 1);
    R_blocks = cell(U, 1);
    energy_term = 0;

    % forward pass: compute S_u and log|S_u|
    for u = 1:U
        Hu = H_blocks{u};
        Bu = B_blocks{u};
        Ru = Bu * Bu';
        R_blocks{u} = 0.5 * (Ru + Ru');

        HRH = Hu * Ru * Hu';
        S = S + HRH;
        S = 0.5 * (S + S');

        [L, flag] = chol(S, 'lower');

        if flag ~= 0
            [L, flag] = chol(S +1e-9 * eye_Ly, 'lower');
        end

        if flag ~= 0
            phi = Inf;
            grad_vec = [];
            R_blocks = [];
            return;
        end

        logdet_S(u) = 2 * sum(log(diag(L)));
        Linv = L \ eye_Ly;
        invS{u} = Linv' * Linv;
        energy_term = energy_term + w(u) * sum(real(conj(Bu(:)) .* Bu(:)));
    end

    F = sum(delta .* logdet_S) - real(energy_term);
    phi = -F; % minimize negative

    % backward pass: compute gradients
    grad_blocks = cell(U, 1);
    tail = zeros(Ly, Ly);

    for u = U:-1:1
        tail = tail + delta(u) * invS{u};
        Hu = H_blocks{u};
        Bu = B_blocks{u};
        Lu = size(Bu, 1);

        % gradient of F w.r.t. R_u
        grad_R = Hu' * tail * Hu - w(u) * eye(Lu);
        grad_R = 0.5 * (grad_R + grad_R');

        % gradient of F w.r.t. B_u (chain rule)
        grad_B = 2 * grad_R * Bu;

        % negate for minimization
        grad_blocks{u} = -grad_B;
    end

    grad_vec = pack_blocks(grad_blocks, block_sizes);

end

function dir = lbfgs_direction(grad, S_hist, Y_hist)
    %LBFGS_DIRECTION Compute L-BFGS search direction

    if isempty(S_hist)
        dir = -grad;
        return;
    end

    k = length(S_hist);
    alpha = zeros(k, 1);
    rho = zeros(k, 1);
    q = grad;

    for i = k:-1:1
        s = S_hist{i};
        y = Y_hist{i};
        denom = real(y' * s);

        if denom <= 1e-12
            rho(i) = 0;
            alpha(i) = 0;
            continue;
        end

        rho(i) = 1 / denom;
        alpha(i) = rho(i) * real(s' * q);
        q = q - alpha(i) * y;
    end

    sk = S_hist{end};
    yk = Y_hist{end};
    denom = real(yk' * yk);

    if denom <= 1e-12
        H0 = 1;
    else
        H0 = real(sk' * yk) / denom;
    end

    r = H0 * q;

    for i = 1:k

        if rho(i) == 0
            continue;
        end

        y = Y_hist{i};
        s = S_hist{i};
        beta_val = rho(i) * real(y' * r);
        r = r + s * (alpha(i) - beta_val);
    end

    dir = -r;

end

function vec = pack_blocks(blocks, block_sizes)
    %PACK_BLOCKS Pack cell array of matrices into a vector

    U = length(block_sizes);
    total = 0;

    for u = 1:U
        total = total + 2 * block_sizes(u) ^ 2;
    end

    vec = zeros(total, 1);
    offset = 0;

    for u = 1:U
        B = blocks{u};
        len = numel(B);
        vec(offset + (1:len)) = real(B(:));
        vec(offset + len + (1:len)) = imag(B(:));
        offset = offset + 2 * len;
    end

end

function blocks = unpack_blocks(vec, block_sizes)
    %UNPACK_BLOCKS Unpack vector into cell array of matrices

    U = length(block_sizes);
    blocks = cell(U, 1);
    offset = 0;

    for u = 1:U
        Lu = block_sizes(u);
        len = Lu ^ 2;
        re = vec(offset + (1:len));
        im = vec(offset + len + (1:len));
        blocks{u} = reshape(re + 1j * im, Lu, Lu);
        offset = offset + 2 * len;
    end

end

function val = logdet_spd(M)
    %LOGDET_SPD Stable log-determinant for SPD matrix

    M = 0.5 * (M + M');
    [L, flag] = chol(M, 'lower');

    if flag ~= 0
        M = M +1e-9 * eye(size(M, 1));
        [L, flag] = chol(M, 'lower');
    end

    if flag ~= 0
        val = -Inf;
        return;
    end

    val = 2 * sum(log(diag(L)));

end

File: minpmac/cpp/solve_tone.cpp
/**
 * @file solve_tone.cpp
 * @brief Per-tone L-BFGS optimization implementation for minPMAC
 */

#include "solve_tone.hpp"

#include <algorithm>
#include <cmath>
#include <numeric>

namespace minpmac {

namespace {

void pack_blocks(const std::vector<MatrixXcd>& blocks,
                 const std::vector<int>& block_sizes, VectorXd& vec) {
    int offset = 0;
    for (size_t u = 0; u < blocks.size(); ++u) {
        const auto& B = blocks[u];
        const int n = B.size();
        for (int i = 0; i < n; ++i) {
            vec(offset + i) = std::real(B.data()[i]);
            vec(offset + n + i) = std::imag(B.data()[i]);
        }
        offset += 2 * n;
    }
}

void unpack_blocks(const VectorXd& vec, const std::vector<int>& block_sizes,
                   std::vector<MatrixXcd>& blocks) {
    int offset = 0;
    for (size_t u = 0; u < block_sizes.size(); ++u) {
        const int Lu = block_sizes[u];
        const int n = Lu * Lu;
        blocks[u].resize(Lu, Lu);
        for (int i = 0; i < n; ++i) {
            blocks[u].data()[i] = Complex(vec(offset + i), vec(offset + n + i));
        }
        offset += 2 * n;
    }
}

/**
 * @brief Evaluate objective and gradient for L-BFGS
 *
 * Objective: F = sum_k alpha_k * log|S_k| - sum_u w_u * ||B_u||^2
 * where S_k = I + sum_{j=1}^k H_j * B_j * B_j' * H_j'
 *
 * We minimize phi = -F
 */
bool eval_objective(const VectorXd& x, const std::vector<MatrixXcd>& H_blocks,
                    const VectorXd& alpha, const VectorXd& sw,
                    const std::vector<int>& block_sizes, int Ly,
                    ToneWorkspace& ws, double& phi, VectorXd& grad_vec) {
    const int U = H_blocks.size();

    unpack_blocks(x, block_sizes, ws.B_blocks);

    ws.S = MatrixXcd::Identity(Ly, Ly);
    double energy_term = 0.0;

    for (int u = 0; u < U; ++u) {
        const auto& Hu = H_blocks[u];
        const auto& Bu = ws.B_blocks[u];

        ws.R_blocks[u] = Bu * Bu.adjoint();
        ws.R_blocks[u] = make_hermitian(ws.R_blocks[u]);

        ws.HRH.noalias() = Hu * ws.R_blocks[u] * Hu.adjoint();
        ws.S += ws.HRH;
        ws.S = make_hermitian(ws.S);

        double logdet;
        if (!cholesky_with_inv(ws.S, ws.L, ws.invS[u], logdet)) {
            phi = std::numeric_limits<double>::infinity();
            return false;
        }

        ws.logdet_S(u) = logdet;

        energy_term += sw(u) * ws.B_blocks[u].squaredNorm();
    }

    double F = alpha.dot(ws.logdet_S) - energy_term;
    phi = -F;

    ws.tail.setZero(Ly, Ly);

    int offset = 0;
    for (int u = 0; u < U; ++u) {
        offset += 2 * block_sizes[u] * block_sizes[u];
    }

    for (int u = U - 1; u >= 0; --u) {
        const int Lu = block_sizes[u];
        const int n = Lu * Lu;
        offset -= 2 * n;

        ws.tail += alpha(u) * ws.invS[u];

        const auto& Hu = H_blocks[u];
        const auto& Bu = ws.B_blocks[u];

        ws.grad_R.noalias() = Hu.adjoint() * ws.tail * Hu;
        ws.grad_R -= sw(u) * MatrixXcd::Identity(Lu, Lu);
        ws.grad_R = make_hermitian(ws.grad_R);

        ws.grad_B.noalias() = 2.0 * ws.grad_R * Bu;

        for (int i = 0; i < n; ++i) {
            grad_vec(offset + i) = -std::real(ws.grad_B.data()[i]);
            grad_vec(offset + n + i) = -std::imag(ws.grad_B.data()[i]);
        }
    }

    return true;
}

/**
 * @brief Compute L-BFGS search direction
 */
void lbfgs_direction(const VectorXd& grad, const std::vector<VectorXd>& S_hist,
                     const std::vector<VectorXd>& Y_hist, VectorXd& alpha_ws,
                     VectorXd& rho_ws, VectorXd& q_ws, VectorXd& dir) {
    const int k = S_hist.size();

    if (k == 0) {
        dir = -grad;
        return;
    }

    q_ws = grad;

    for (int i = k - 1; i >= 0; --i) {
        const double denom = Y_hist[i].dot(S_hist[i]);
        if (denom <= 1e-12) {
            rho_ws(i) = 0.0;
            alpha_ws(i) = 0.0;
            continue;
        }
        rho_ws(i) = 1.0 / denom;
        alpha_ws(i) = rho_ws(i) * S_hist[i].dot(q_ws);
        q_ws -= alpha_ws(i) * Y_hist[i];
    }

    const double yk_yk = Y_hist.back().squaredNorm();
    double H0 = 1.0;
    if (yk_yk > 1e-12) {
        H0 = S_hist.back().dot(Y_hist.back()) / yk_yk;
    }

    dir = H0 * q_ws;

    for (int i = 0; i < k; ++i) {
        if (rho_ws(i) == 0.0) continue;
        const double beta = rho_ws(i) * Y_hist[i].dot(dir);
        dir += S_hist[i] * (alpha_ws(i) - beta);
    }

    dir = -dir;
}

}  // namespace

ToneResult solve_tone(const MatrixXcd& Hn, const VectorXd& theta,
                      const VectorXd& w, const std::vector<int>& Lxu,
                      const std::vector<int>& idx_start,
                      const std::vector<int>& idx_end, LBFGSState& state_in,
                      ToneWorkspace& ws) {
    const int Ly = Hn.rows();
    const int U = Lxu.size();

    std::vector<int> order(U);
    std::iota(order.begin(), order.end(), 0);
    std::sort(order.begin(), order.end(),
              [&](int a, int b) { return theta(a) > theta(b); });

    VectorXd stheta(U), sw(U);
    std::vector<int> sLxu(U);
    for (int i = 0; i < U; ++i) {
        stheta(i) = theta(order[i]);
        sw(i) = w(order[i]);
        sLxu[i] = Lxu[order[i]];
    }

    ws.H_blocks.resize(U);
    for (int u = 0; u < U; ++u) {
        int orig = order[u];
        ws.H_blocks[u] =
            Hn.middleCols(idx_start[orig], idx_end[orig] - idx_start[orig] + 1);
    }

    VectorXd alpha(U);
    for (int u = 0; u < U - 1; ++u) {
        alpha(u) = 0.5 * (stheta(u) - stheta(u + 1));
    }
    alpha(U - 1) = 0.5 * stheta(U - 1);

    bool state_valid = state_in.valid && state_in.order == order &&
                       state_in.block_sizes == Lxu;

    int total_params = 0;
    for (int u = 0; u < U; ++u) {
        total_params += 2 * sLxu[u] * sLxu[u];
    }

    ws.B_blocks.resize(U);
    ws.R_blocks.resize(U);

    if (state_valid) {
        ws.B_blocks = state_in.B_blocks;
    } else {
        for (int u = 0; u < U; ++u) {
            ws.B_blocks[u] =
                std::sqrt(1e-6) * MatrixXcd::Identity(sLxu[u], sLxu[u]);
        }
        state_in.S_hist.clear();
        state_in.Y_hist.clear();
    }

    ws.x.resize(total_params);
    ws.x_new.resize(total_params);
    ws.x_trial.resize(total_params);
    ws.grad.resize(total_params);
    ws.grad_new.resize(total_params);
    ws.grad_trial.resize(total_params);
    ws.dir.resize(total_params);
    ws.logdet_S.resize(U);
    ws.invS.resize(U);
    for (int u = 0; u < U; ++u) {
        ws.invS[u].resize(Ly, Ly);
    }

    LBFGSConfig cfg;
    cfg.max_iter = 100;
    cfg.m_hist = 8;
    cfg.tol_grad = 1e-5;

    pack_blocks(ws.B_blocks, sLxu, ws.x);

    double phi;
    if (!eval_objective(ws.x, ws.H_blocks, alpha, sw, sLxu, Ly, ws, phi,
                        ws.grad)) {
        for (int u = 0; u < U; ++u) {
            ws.B_blocks[u] =
                std::sqrt(1e-6) * MatrixXcd::Identity(sLxu[u], sLxu[u]);
        }
        pack_blocks(ws.B_blocks, sLxu, ws.x);
        state_in.S_hist.clear();
        state_in.Y_hist.clear();
        eval_objective(ws.x, ws.H_blocks, alpha, sw, sLxu, Ly, ws, phi,
                       ws.grad);
    }

    ws.alpha.resize(cfg.m_hist);
    ws.rho.resize(cfg.m_hist);
    ws.q.resize(total_params);

    double grad_norm_sq = ws.grad.squaredNorm();
    const double tol_grad_sq = cfg.tol_grad * cfg.tol_grad;

    for (int iter = 0; iter < cfg.max_iter; ++iter) {
        if (grad_norm_sq < tol_grad_sq) {
            break;
        }

        lbfgs_direction(ws.grad, state_in.S_hist, state_in.Y_hist, ws.alpha,
                        ws.rho, ws.q, ws.dir);

        double dir_deriv = ws.grad.dot(ws.dir);

        if (dir_deriv >= 0) {
            ws.dir = -ws.grad;
            dir_deriv = ws.grad.dot(ws.dir);
            state_in.S_hist.clear();
            state_in.Y_hist.clear();
        }

        double step = cfg.step_init;
        bool accepted = false;
        double phi_new = phi;

        for (int ls = 0; ls < cfg.max_linesearch; ++ls) {
            ws.x_trial = ws.x + step * ws.dir;

            double phi_trial;
            if (eval_objective(ws.x_trial, ws.H_blocks, alpha, sw, sLxu, Ly, ws,
                               phi_trial, ws.grad_trial)) {
                if (phi_trial <= phi + cfg.armijo_c * step * dir_deriv) {
                    accepted = true;
                    ws.x_new = ws.x_trial;
                    phi_new = phi_trial;
                    ws.grad_new = ws.grad_trial;
                    break;
                }
            }

            step *= cfg.beta;
            if (step < cfg.step_min) break;
        }

        if (!accepted) {
            ws.dir = -ws.grad;
            dir_deriv = ws.grad.dot(ws.dir);
            step = cfg.step_init;

            for (int ls = 0; ls < cfg.max_linesearch; ++ls) {
                ws.x_trial = ws.x + step * ws.dir;

                double phi_trial;
                if (eval_objective(ws.x_trial, ws.H_blocks, alpha, sw, sLxu, Ly,
                                   ws, phi_trial, ws.grad_trial)) {
                    if (phi_trial <= phi + cfg.armijo_c * step * dir_deriv) {
                        accepted = true;
                        ws.x_new = ws.x_trial;
                        phi_new = phi_trial;
                        ws.grad_new = ws.grad_trial;
                        break;
                    }
                }

                step *= cfg.beta;
                if (step < cfg.step_min) break;
            }

            if (!accepted) break;

            state_in.S_hist.clear();
            state_in.Y_hist.clear();
        }

        VectorXd s = ws.x_new - ws.x;
        VectorXd y = ws.grad_new - ws.grad;
        double sy = s.dot(y);

        if (sy > 1e-12) {
            if (static_cast<int>(state_in.S_hist.size()) >= cfg.m_hist) {
                state_in.S_hist.erase(state_in.S_hist.begin());
                state_in.Y_hist.erase(state_in.Y_hist.begin());
            }
            state_in.S_hist.push_back(s);
            state_in.Y_hist.push_back(y);
        } else {
            state_in.S_hist.clear();
            state_in.Y_hist.clear();
        }

        ws.x = ws.x_new;
        phi = phi_new;
        ws.grad = ws.grad_new;
        grad_norm_sq = ws.grad.squaredNorm();
    }

    unpack_blocks(ws.x, sLxu, ws.B_blocks);
    for (int u = 0; u < U; ++u) {
        ws.R_blocks[u] = ws.B_blocks[u] * ws.B_blocks[u].adjoint();
        ws.R_blocks[u] = make_hermitian(ws.R_blocks[u]);
    }

    VectorXd b_sorted(U);
    MatrixXcd S_curr = MatrixXcd::Identity(Ly, Ly);
    double logdet_prev = 0.0;

    for (int u = 0; u < U; ++u) {
        S_curr += ws.H_blocks[u] * ws.R_blocks[u] * ws.H_blocks[u].adjoint();
        S_curr = make_hermitian(S_curr);
        double logdet_curr = logdet_spd(S_curr);
        b_sorted(u) = 0.5 * (logdet_curr - logdet_prev);
        logdet_prev = logdet_curr;
    }

    VectorXd E_sorted(U);
    for (int u = 0; u < U; ++u) {
        E_sorted(u) = std::real(ws.R_blocks[u].trace());
    }

    ToneResult result;
    result.b.resize(U);
    result.E.resize(U);
    result.Rxx.resize(U);

    for (int u = 0; u < U; ++u) {
        int orig = order[u];
        result.b(orig) = b_sorted(u);
        result.E(orig) = E_sorted(u);
        result.Rxx[orig] = ws.R_blocks[u];
    }

    result.f = theta.dot(result.b) - w.dot(result.E);

    state_in.B_blocks = ws.B_blocks;
    state_in.order = order;
    state_in.block_sizes = Lxu;
    state_in.valid = true;

    return result;
}

}  // namespace minpmac
